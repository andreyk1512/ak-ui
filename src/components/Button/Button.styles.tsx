import { css } from '@emotion/react';
import * as styles from '../../styles';
import { ButtonProps } from './Button.types';

export const loaderStyles = ({ hidden }: { hidden: boolean }) => css`
  height: 12px;
  width: 12px;
  color: currentColor;
  display: inline-block;
  border: 1px solid;
  border-radius: 50%;
  margin-inline-end: 8px;
  border-top-color: transparent;
  animation: rotate 1s linear infinite;
  transition: all 0.2s;

  ${hidden &&
  css`
    scale: 0;
    width: 0;
    margin-inline-end: 0;
  `};

  @keyframes rotate {
    0% {
      transform: rotate(0);
    }
    100% {
      transform: rotate(360deg);
    }
  }
`;

export const buttonStyles = ({ size, type, ghost }: Pick<ButtonProps, 'size' | 'type' | 'ghost'>) => css`
  display: flex;
  justify-content: center;
  align-items: center;
  white-space: nowrap;
  box-sizing: border-box;
  width: auto;
  height: 45px;
  cursor: pointer;
  user-select: none;
  text-align: center;
  line-height: 100%;
  padding: 0 15px;
  border-radius: 5px;
  font-style: normal;
  font-size: ${styles.fontSize.regular};
  background-color: transparent;
  box-shadow: none;
  border: 1px solid ${styles.color.lightSeaGreen};
  color: ${styles.color.yankeesBlue};
  transition: all 0.2s;

  &:hover {
    background-color: ${styles.color.lightSeaGreen};
    border-color: ${styles.color.lightSeaGreen};
    color: ${styles.color.white};
  }

  &:disabled {
    opacity: 0.65;
    cursor: not-allowed;
  }

  ${size === 'small' &&
  css`
    padding: 6px 15px;
    border-radius: 3px;
    height: 27px;
  `}

  ${type === 'default' &&
  css`
    ${ghost &&
    css`
      color: ${styles.color.white};
    `}
  `}

  ${type === 'primary' &&
  css`
    color: ${styles.color.white};
    background-color: ${styles.color.lightSeaGreen};
    border: 1px solid ${styles.color.lightSeaGreen};
    box-shadow: 0 4px 4px 0 rgba(0, 0, 0, 0.25);

    &:hover {
      background-color: ${styles.color.greenCrayola};
      border-color: ${styles.color.greenCrayola};
    }

    ${ghost &&
    css`
      color: ${styles.color.white};
    `}
  `}

    ${type === 'secondary' &&
  css`
    color: ${styles.color.white};
    background-color: ${styles.color.tuftsBlue};
    border: 1px solid ${styles.color.tuftsBlue};

    &:hover {
      background-color: ${styles.color.greenBlue};
      border-color: ${styles.color.greenBlue};
    }
  `}

  ${ghost &&
  css`
    background-color: transparent;
  `}
`;
