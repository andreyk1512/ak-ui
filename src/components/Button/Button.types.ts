import { MouseEventHandler, ReactNode } from 'react';

export type ButtonProps = {
  children: ReactNode;
  size?: 'default' | 'small';
  type?: 'default' | 'primary' | 'secondary';
  disabled?: boolean;
  loading?: boolean;
  ghost?: boolean;
  onClick?: MouseEventHandler<HTMLButtonElement>;
  extraClass?: string;
  htmlType?: HTMLButtonElement['type'];
};
