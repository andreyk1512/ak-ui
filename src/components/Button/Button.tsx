import { buttonStyles, loaderStyles } from './Button.styles';
import { ButtonProps } from './Button.types';

export const Button = ({
  children,
  size = 'default',
  type = 'default',
  disabled,
  loading,
  ghost,
  onClick,
  extraClass,
  htmlType = 'button',
}: ButtonProps) => {
  return (
    <button
      css={buttonStyles({ size, type, ghost })}
      disabled={disabled}
      onClick={onClick}
      className={extraClass}
      type={htmlType}
    >
      {loading ? <span role="loader" css={loaderStyles} /> : null}
      {children}
    </button>
  );
};
