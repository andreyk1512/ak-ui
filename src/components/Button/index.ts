export * from './Button';
export * from './Button.types';

import { Link } from './components';
import { Button as OriginButton } from './Button';

export type ButtonProps = typeof OriginButton & {
  Link: typeof Link;
};

export const Button = OriginButton as ButtonProps;
Button.Link = Link;
