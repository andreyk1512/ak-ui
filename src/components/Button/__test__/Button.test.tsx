import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { Button } from '../Button';

describe('Button', () => {
  it('should render the button text', () => {
    const { container } = render(<Button>Click Me</Button>);
    const button = screen.getByRole('button');

    expect(button).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should disable the button if the disabled prop is true', () => {
    const onClickHandler = jest.fn();
    render(<Button disabled onClick={onClickHandler}>Click Me</Button>);
    const button = screen.getByRole('button');
    button.click();

    expect(button).toBeDisabled();
    expect(onClickHandler).not.toHaveBeenCalled();
  });

  it('should show the button loader if the loading prop is true', () => {
    render(<Button loading>Click Me</Button>);
    const loader = screen.getByRole('button');

    expect(loader).toBeInTheDocument();
  });

  it('should fire the onClick event when the button is clicked', () => {
    const onClickHandler = jest.fn();
    render(<Button onClick={onClickHandler}>Click Me</Button>);
    const button = screen.getByRole('button');
    button.click();

    expect(onClickHandler).toBeCalled();
  });
});
