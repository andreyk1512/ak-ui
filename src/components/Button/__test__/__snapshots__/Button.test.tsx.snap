// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render the button text 1`] = `
<button
  class="css-4cc993-Button"
  type="button"
>
  Click Me
</button>
`;
